commons/src/main/scala/za/co/absa/spline/common/rest/RESTClient.scala

```
@@ -0,0 +1,85 @@
+import za.co.absa.commons.lang.ARM
+import za.co.absa.commons.lang.ARM.managed
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class RESTClient(uri: URI, maybeCredentials: Option[Credentials]) {
+  @throws[HttpStatusException]
+  def get(path: String): Future[String] = execHttp {
+    baseUri => new HttpGet(s"$baseUri/$path")
+  }
+
+  @throws[HttpStatusException]
+  def delete(path: String)(implicit ec: ExecutionContext): Future[Unit] = execHttp {
+    baseUri => new HttpDelete(s"$baseUri/$path")
+  }.map(_ => {})
+
+  @throws[HttpStatusException]
+  def post(path: String, body: String)(implicit ec: ExecutionContext): Future[Unit] = execHttp {
+    baseUri =>
+      new HttpPost(s"$baseUri/$path") {
+        setEntity(new StringEntity(body))
+      }
+  }.map(_ => {})
+
+  @throws[HttpStatusException]
+  private def execHttp(method: URI => HttpRequestBase): Future[String] = {
+    val request = {
+      val req = method(uri)
+      maybeCredentials.foreach(credentials => {
+        val authHeader = new BasicScheme().authenticate(credentials, req, null)
+        req.addHeader(authHeader)
+      })
+      req
+    }
+
+    val (respStatusLine, respBody) =
+      for {
+        httpClient <- managed(HttpClients.createDefault)
+        response <- managed(httpClient.execute(request))
+      } yield {
+        val maybeBody = Option(response.getEntity)
+          .map(e => {
+            val encoding = Option(e.getContentEncoding).map(_.getValue).getOrElse("UTF-8")
+            ARM.using(e.getContent) {
+              inputStream =>
+                IOUtils.toString(inputStream, encoding)
+            }
+          })
+        (response.getStatusLine, maybeBody.orNull)
+      }
+
+    respStatusLine.getStatusCode match {
+      case 200 | 201 | 204 =>
+        Future.successful(respBody)
+      case _ =>
+        throw new HttpStatusException(respStatusLine.getStatusCode, s"ArangoDB response: $respStatusLine. $respBody")
```

Why not use Future.failed() instead of throwing?

Or opposite question. Is the Future actually useful here? We could return `respBody` immediately.

---

Good point. Actually I'd prefer to keep async nature of the methods. Even though the current implementation uses a blocking Http client, wrapping it with a future makes it easy to change the impl anytime.

Would something like this read better?
```scala
    Future.fromTry(Try(respStatusLine.getStatusCode match {
      case 200 | 201 | 204 =>
        respBody
      case _ =>
        throw new HttpStatusException(respStatusLine.getStatusCode, s"ArangoDB response: $respStatusLine. $respBody")
    }))
```

---

`Future.fromTry(Try` can even be placed on the beginning of the method

---

This would be the easiest to read for me:
```
    respStatusLine.getStatusCode match {
      case 200 | 201 | 204 =>
        Future.successful(respBody)
      case statusCode =>
        Future.failed(new HttpStatusException(statusCode, s"ArangoDB response: $respStatusLine. $respBody"))
    }
```

But wrapping the whole method in `Future.fromTry(Try` would catch all exceptions happening there, but I don't know whether that is something we want.

Both options are ok for me.

---

I'd say if a method returns a `Future` then it shouldn't throw. It often leads to inconsistency and difficulties in error handling.

---

done